evm/src/cpu/jumps.rs

```diff
@@ -0,0 +1,376 @@
+use once_cell::sync::Lazy;
+use plonky2::field::extension::Extendable;
+use plonky2::field::packed::PackedField;
+use plonky2::field::types::Field;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::cpu::columns::CpuColumnsView;
+use crate::cpu::kernel::aggregator::KERNEL;
+
+static INVALID_DST_HANDLER_ADDR: Lazy<usize> =
+    Lazy::new(|| KERNEL.global_labels["fault_exception"]);
+
+pub fn eval_packed_exit_kernel<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    nv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let jumps_lv = lv.general.jumps();
+
+    // If we are executing `EXIT_KERNEL` then we simply restore the program counter and kernel mode
+    // flag. The top 6 (32-bit) limbs are ignored (this is not part of the spec, but we trust the
+    // kernel to set them to zero).
+    yield_constr.constraint_transition(
+        lv.is_cpu_cycle * lv.is_exit_kernel * (jumps_lv.input0[0] - nv.program_counter),
+    );
+    yield_constr.constraint_transition(
+        lv.is_cpu_cycle * lv.is_exit_kernel * (jumps_lv.input0[1] - nv.is_kernel_mode),
+    );
+}
+
+pub fn eval_ext_circuit_exit_kernel<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut plonky2::plonk::circuit_builder::CircuitBuilder<F, D>,
+    lv: &CpuColumnsView<ExtensionTarget<D>>,
+    nv: &CpuColumnsView<ExtensionTarget<D>>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    let jumps_lv = lv.general.jumps();
+    let filter = builder.mul_extension(lv.is_cpu_cycle, lv.is_exit_kernel);
+
+    // If we are executing `EXIT_KERNEL` then we simply restore the program counter and kernel mode
+    // flag. The top 6 (32-bit) limbs are ignored (this is not part of the spec, but we trust the
+    // kernel to set them to zero).
+
+    let pc_constr = builder.sub_extension(jumps_lv.input0[0], nv.program_counter);
+    let pc_constr = builder.mul_extension(filter, pc_constr);
+    yield_constr.constraint_transition(builder, pc_constr);
+
+    let kernel_constr = builder.sub_extension(jumps_lv.input0[1], nv.is_kernel_mode);
+    let kernel_constr = builder.mul_extension(filter, kernel_constr);
+    yield_constr.constraint_transition(builder, kernel_constr);
+}
+
+pub fn eval_packed_jump_jumpi<P: PackedField>(
+    lv: &CpuColumnsView<P>,
+    nv: &CpuColumnsView<P>,
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let jumps_lv = lv.general.jumps();
+    let filter = lv.is_jump + lv.is_jumpi; // `JUMP` or `JUMPI`
```

Comment:
Do we need to filter by `is_cpu_cycle` as well?

(If the degree increase is problematic I think we could always make `is_jump` dedicated. I think the plan was to share them with kernel bootloading rows, but we could do without.)

Reply:
Ah yes, this is a documentation bug. I intended to make `is_jump` and `is_jumpi` dedicated due to degree constraints. I’ll make a note of that.